torch/lib/THD/base/data_channels/DataChannelGloo.hpp

```diff
@@ -17,17 +23,190 @@ namespace std {
 
 template<>
 struct hash<std::tuple<::thd::DataOperation, THDGroup>> {
-  std::size_t operator()(const std::tuple<::thd::DataOperation,
-                         THDGroup>& k) const {
-    return hash<::thd::DataOperation>()(std::get<0>(k)) ^
-        hash<THDGroup>()(std::get<1>(k));
+  std::size_t operator()(const std::tuple<::thd::DataOperation, THDGroup>& k) const {
+    return (
+      hash<::thd::DataOperation>()(std::get<0>(k)) ^
+      hash<THDGroup>()(std::get<1>(k))
+    );
+  }
+};
+
+template<>
+struct hash<std::tuple<::thd::DataOperation, THDGroup, std::size_t,
+                       std::size_t, THDReduceOp, thd::rank_type>> {
+  std::size_t operator()(const std::tuple<::thd::DataOperation, THDGroup,
+                                          std::size_t, std::size_t, THDReduceOp,
+                                          thd::rank_type>& k) const {
+    return (
+      hash<::thd::DataOperation>()(std::get<0>(k)) ^
+      hash<THDGroup>()(std::get<1>(k)) ^
+      hash<std::size_t>()(std::get<2>(k)) ^
+      hash<std::size_t>()(std::get<3>(k)) ^
+      hash<THDReduceOp>()(std::get<4>(k)) ^
+      hash<thd::rank_type>()(std::get<5>(k))
```

Comment:
It's not really a good way of hashing multiple values. Look for implementation of `boost::combine_hash`.